File: 02. Estructurales/02. Bridge/01. bridge.js
/***

Imaginemos que estamos construyendo un sistema de reproducción de medios que puede 
reproducir diferentes tipos de archivos multimedia, como música y videos. 
Además, queremos que nuestro sistema sea compatible con diferentes reproductores, 
como reproductores de audio y reproductores de video.

Aquí es donde el patrón Bridge puede ser útil. 
Podemos utilizar el patrón Bridge para separar las abstracciones de los archivos multimedia
de las implementaciones de los reproductores.

*/


// Comencemos definiendo nuestras implementaciones de reproductores:

// Implementación de reproductor de audio
class AudioPlayer {
    play() {
        console.log("Reproduciendo audio...");
    }
    pause() {
        console.log("Pausando audio...");
    }
    stop() {
        console.log("Deteniendo audio...");
    }
}

// Implementación de reproductor de video
class VideoPlayer {
    play() {
        console.log("Reproduciendo video...");
    }
    pause() {
        console.log("Pausando video...");
    }
    stop() {
        console.log("Deteniendo video...");
    }
}

/***

A continuación, definimos nuestra clase abstracta de archivo multimedia 
que actuará como el puente entre las abstracciones de los archivos multimedia 
y las implementaciones de los reproductores:

*/

class MediaFile {
    constructor(player) {
        this.player = player;
    }

    play() {
        this.player.play();
    }

    pause() {
        this.player.pause();
    }

    stop() {
        this.player.stop();
    }
}

/***

Finalmente, podemos utilizar el patrón Bridge para reproducir 
diferentes tipos de archivos multimedia con diferentes reproductores:

*/

// Crear una instancia de archivo de música con reproductor de audio
const audioFile = new MediaFile(new AudioPlayer());
audioFile.play();

// Crear una instancia de archivo de video con reproductor de video
const videoFile = new MediaFile(new VideoPlayer());
videoFile.play();


/***

En este ejemplo, el patrón Bridge nos permite separar las abstracciones 
de los archivos multimedia de las implementaciones de los reproductores. 
Esto nos permite cambiar fácilmente los reproductores para diferentes tipos 
de archivos multimedia sin afectar el código que reproduce los archivos.

*/

File: 02. Estructurales/02. Bridge/01. bridge.py
# Implementación de reproductor de audio
class AudioPlayer:
    def play(self):
        print("Reproduciendo audio...")

    def pause(self):
        print("Pausando audio...")

    def stop(self):
        print("Deteniendo audio...")

# Implementación de reproductor de video
class VideoPlayer:
    def play(self):
        print("Reproduciendo video...")

    def pause(self):
        print("Pausando video...")

    def stop(self):
        print("Deteniendo video...")

# Definimos nuestra clase abstracta de archivo multimedia
class MediaFile:
    def __init__(self, player):
        self.player = player

    def play(self):
        self.player.play()

    def pause(self):
        self.player.pause()

    def stop(self):
        self.player.stop()

# Utilizamos el patrón Bridge para reproducir diferentes tipos de archivos multimedia con diferentes reproductores

# Crear una instancia de archivo de música con reproductor de audio
audio_file = MediaFile(AudioPlayer())
audio_file.play()

# Crear una instancia de archivo de video con reproductor de video
video_file = MediaFile(VideoPlayer())
video_file.play()